src/Dialect/ONNX/ONNXDimAnalysis.cpp

```diff
@@ -68,21 +68,30 @@ static bool areOverlapping(
 static std::optional<DimAnalysis::DimT> insertDimWhenUseful(const Value tensor,
     const uint64_t dimIndex, DimAnalysis::DimSetT &sameDims) {
   auto tensorType = cast<ShapedType>(tensor.getType());
+  uint64_t axis = dimIndex;
 
   bool okToInsert = false;
   if (tensor.isa<BlockArgument>()) {
     okToInsert = true;
   } else {
     Operation *op = tensor.getDefiningOp();
-    if (isa<ONNXConstantOp, ONNXCastOp, ONNXDimOp>(op) ||
-        tensorType.isDynamicDim(dimIndex))
+    // A constant of -1 to define a dynamic value, e.g. -1 in Reshape means a
+    // dynamic dimension computed from the other dimensions. It's a contant, no
+    // need to insert it.
+    if (isa<ONNXConstantOp>(op))
+      okToInsert = false;
+    else if (auto dimOp = dyn_cast<ONNXDimOp>(op)) {
+      // The correct axis is from ONNXDimOp.
+      axis = dimOp.getAxis();
+      okToInsert = true;
+    } else if (isa<ONNXCastOp>(op) || tensorType.isDynamicDim(axis))
```

Comment:
I am trying the understand the difference: for DimOp, the axis is the for the input, while for other op, the axis is for the output. The outputIE of other ops is somehow connected to its inputIE by shape analysis, but not for DimOp? Special case is always questionable to me.

Reply:
Tong, the relationship is that inputIE is defined by DimOp.